### ISSUE TITLE
[Summer 2026 LFX] 12-Week Implementation Plan

### ISSUE THREAD

**Issue by arunthakur009** [merledu/rv-sparse#5]
<html><head></head><body><h1>[Summer 2026 LFX] 12-Week Implementation Plan</h1>
<h2>rv-sparse: Open-source RISC-V Vector Accelerated Sparse Linear Algebra Library</h2>
<p><strong>Applicant:</strong> Arun Thakur
<strong>Project:</strong> Broadening the RISC-V High Precision Code Base and Reach
<strong>Mentors:</strong> Talha Ahmed, Shayan Baig
<strong>Term:</strong> Summer 2026</p>
<hr>
<h2>About Me</h2>
<p>I am a second-year Engineering Physics student, with a strong mathematical foundation in linear algebra, numerical methods, and computational physics — all directly relevant to sparse linear algebra work.</p>
<p>My hands-on experience with RISC-V began as a <strong>RISC-V Backend Contributor to Intel OpenVINO</strong>, where I implemented JIT emitter optimizations for RISC-V64 with RVV1.0 vector extensions in C++, contributing low-level code generation for ML inference on RISC-V hardware. This work gave me direct exposure to instruction set architecture, vector intrinsics, and cross-architecture code emission — exactly the skill set required for <code>rv-sparse</code>.</p>
<p>Concurrently, as an <strong>ML Infrastructure Contributor to Keras</strong>, I developed production-grade NumPy runtime operators in Python, implementing numerical computing kernels with focus on precision, performance, and cross-platform compatibility. This deepened my understanding of how operator-level design decisions affect throughput and accuracy at scale.</p>
<p>My technical stack includes C, C++, Python, and Bash scripting, with familiarity across PyTorch, OpenVINO, NumPy, SciPy, and CMake-based build systems on Linux. I am comfortable with Git workflows, continuous integration, and performance profiling, all of which are essential for an open-source library contribution of this scope.</p>
<hr>
<h2>My Understanding of the Project</h2>
<p>Sparse linear algebra underpins a wide range of modern computing workloads — from scientific simulations and graph analytics to pruned neural networks where sparsity is deliberately engineered to reduce computation and memory footprint. The key challenge is that conventional dense BLAS routines waste cycles operating on zeros, while existing hardware accelerators for sparse operations are not universally available in general-purpose systems.</p>
<p>Within the RISC-V ecosystem, <strong>OpenBLAS</strong> provides highly optimized dense linear algebra, but comparable support for <strong>sparse matrix multiplication using RVV (RISC-V Vector Extension)</strong> is largely absent. This project addresses that gap directly.</p>
<p>The two canonical sparse storage formats at the heart of this project are:</p>
<ul>
<li><strong>CSR (Compressed Sparse Row):</strong> Efficient for row-wise access and SpMV (sparse matrix–vector multiply). Stores row pointers, column indices, and values.</li>
<li><strong>CSC (Compressed Sparse Column):</strong> The column-wise counterpart; useful for certain factorization and transposition-heavy workloads.</li>
</ul>
<p>The core vectorization challenge in sparse computation is <strong>irregular memory access</strong> — the non-zero positions are not contiguous, so naïve vectorization is ineffective. The RISC-V Vector Extension (RVV 1.0) provides powerful primitives for this:</p>
<ul>
<li><strong>Masked vector loads/stores</strong> to conditionally process only non-zero lanes</li>
<li><strong>Indexed (scatter/gather) memory operations</strong> (<code>vluxei</code>, <code>vsuxei</code>) for non-contiguous access patterns</li>
<li><strong>Vector length agnosticism (VLEN)</strong> allowing the same code to run efficiently across different RVV implementations</li>
<li><strong>Reduction operations</strong> (<code>vredsum</code>) for efficient dot product accumulation</li>
</ul>
<p>The expected outcome is a clean, well-documented C library exposing a stable API for sparse matmul operations (SpMM and SpMV), with vectorized inner loops written using RVV intrinsics, and a test/benchmark suite that validates correctness and measures performance against scalar baselines.</p>
<hr>
<h2>12-Week Implementation Plan</h2>
<h3>Phase 1 — Foundation and Environment Setup (Weeks 1–3)</h3>
<h4>Week 1: Environment, Toolchain, and Repository Bootstrap</h4>
<p><strong>Goals:</strong> Establish a fully functional RISC-V development environment and define project structure.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Set up cross-compilation toolchain for RISC-V64 with RVV 1.0 support (<code>riscv64-unknown-linux-gnu-gcc</code> with <code>-march=rv64gcv</code>)</li>
<li>Configure QEMU RISC-V user-mode emulation for local testing without physical hardware</li>
<li>Initialize the <code>rv-sparse</code> repository with CMake build system supporting both native and cross-compilation targets</li>
<li>Define directory layout: <code>include/</code>, <code>src/</code>, <code>tests/</code>, <code>benchmarks/</code>, <code>docs/</code></li>
<li>Set up CI pipeline (GitHub Actions) with QEMU-based RISC-V test runner</li>
<li>Write a minimal <code>CONTRIBUTING.md</code> and coding style guide</li>
</ul>
<p><strong>Deliverables:</strong> Functional build system, CI passing a "Hello RVV" smoke test, initial README.</p>
<hr>
<h4>Week 2: CSR and CSC Format Implementation</h4>
<p><strong>Goals:</strong> Implement core sparse storage formats with a clean C API.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Define the primary data structures in <code>include/rv_sparse.h</code>:
<ul>
<li><code>rv_sparse_csr_t</code>: row pointers (<code>rowptr</code>), column indices (<code>colidx</code>), values (<code>vals</code>), dimensions</li>
<li><code>rv_sparse_csc_t</code>: analogous column-major structure</li>
</ul>
</li>
<li>Implement allocation, deallocation, and utility functions: <code>rv_sparse_csr_create</code>, <code>rv_sparse_csr_free</code>, <code>rv_sparse_csr_from_dense</code></li>
<li>Implement format conversion: <code>rv_sparse_csr_to_csc</code>, <code>rv_sparse_csc_to_csr</code></li>
<li>Write unit tests covering edge cases: empty matrix, all-zeros, fully dense, single row/column, large matrices</li>
</ul>
<p><strong>Deliverables:</strong> <code>rv_sparse.h</code> public header, format implementation in <code>src/formats.c</code>, passing unit tests.</p>
<hr>
<h4>Week 3: Scalar Baseline — SpMV and SpMM</h4>
<p><strong>Goals:</strong> Implement correct, portable scalar reference implementations to validate against.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Implement scalar <code>rv_sparse_spmv_csr</code>: CSR sparse matrix–vector multiply (y = A·x)</li>
<li>Implement scalar <code>rv_sparse_spmv_csc</code>: CSC variant</li>
<li>Implement scalar <code>rv_sparse_spmm_csr</code>: CSR sparse matrix–dense matrix multiply (C = A·B)</li>
<li>Validate numerical correctness against NumPy reference outputs for double- and single-precision (fp64, fp32)</li>
<li>Document API contracts: input/output ownership, aliasing rules, error codes</li>
</ul>
<p><strong>Deliverables:</strong> Scalar SpMV and SpMM in <code>src/scalar/</code>, correctness test suite with randomized matrix tests.</p>
<hr>
<h3>Phase 2 — RVV Vectorization (Weeks 4–7)</h3>
<h4>Week 4: RVV Primer and SpMV Vectorization (CSR)</h4>
<p><strong>Goals:</strong> Implement the first RVV-accelerated kernel — CSR SpMV.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Study RVV 1.0 intrinsics documentation; prototype gather-based dot product in isolation</li>
<li>Implement <code>rv_sparse_spmv_csr_rvv</code> using:
<ul>
<li><code>vsetvli</code> for dynamic vector length configuration</li>
<li><code>vluxei64</code> for indexed value loads</li>
<li><code>vfmul</code>/<code>vfmacc</code> for fused multiply-accumulate</li>
<li><code>vfredusum</code> for lane reduction</li>
</ul>
</li>
<li>Verify correctness against scalar baseline with identical inputs</li>
<li>Profile with QEMU performance counters; document vector utilization</li>
</ul>
<p><strong>Deliverables:</strong> <code>src/rvv/spmv_csr_rvv.c</code>, correctness verified, initial performance numbers documented.</p>
<hr>
<h4>Week 5: RVV SpMV — CSC Format and Masked Variants</h4>
<p><strong>Goals:</strong> Extend vectorized SpMV to CSC and add masked execution for irregular sparsity patterns.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Implement <code>rv_sparse_spmv_csc_rvv</code> using column-wise scatter accumulation</li>
<li>Implement masked variant using RVV predicate masks to skip zero-filled vector lanes in ultra-sparse matrices</li>
<li>Benchmark both CSR and CSC variants across varying sparsity levels (1%, 5%, 10%, 30%, 50% density)</li>
<li>Write a sparsity-adaptive dispatch function: selects masked vs. unmasked path based on nnz density</li>
</ul>
<p><strong>Deliverables:</strong> CSC SpMV kernel, masked variant, dispatch logic, benchmark results table.</p>
<hr>
<h4>Week 6: RVV SpMM — Core Vectorization</h4>
<p><strong>Goals:</strong> Implement vectorized sparse matrix–dense matrix multiply (SpMM), the most computationally intensive primitive.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Implement <code>rv_sparse_spmm_csr_rvv</code>:
<ul>
<li>Outer loop over rows of A (sparse); inner loop over columns of B (dense) vectorized with RVV</li>
<li>Use <code>vlse</code> (strided load) for non-contiguous B column access</li>
<li>Accumulate into C with <code>vfmacc</code></li>
</ul>
</li>
<li>Handle non-square and non-power-of-2 dimensions correctly via <code>vsetvli</code> tail handling</li>
<li>Validate against scalar SpMM for fp32 and fp64</li>
</ul>
<p><strong>Deliverables:</strong> <code>src/rvv/spmm_csr_rvv.c</code>, passing precision tests.</p>
<hr>
<h4>Week 7: SpMM Optimization — Blocking and Memory Access Patterns</h4>
<p><strong>Goals:</strong> Improve cache utilization and memory bandwidth efficiency in SpMM.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Implement row-blocking strategy: process multiple rows of A simultaneously to increase register reuse</li>
<li>Implement column-tiling on B to improve L1/L2 cache hit rate</li>
<li>Investigate and implement SELL-C-σ (sliced ELLPACK) as an optional storage format for more regular sparsity patterns</li>
<li>Profile memory access patterns using QEMU and hardware performance model</li>
<li>Document performance tradeoffs between formats and blocking strategies</li>
</ul>
<p><strong>Deliverables:</strong> Blocked SpMM kernel, optional SELL-C-σ support, profiling report.</p>
<hr>
<h3>Phase 3 — API Hardening, Benchmarking, and Documentation (Weeks 8–10)</h3>
<h4>Week 8: API Finalization and Error Handling</h4>
<p><strong>Goals:</strong> Produce a stable, production-quality public API.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Finalize <code>rv_sparse.h</code> with complete Doxygen-annotated API</li>
<li>Implement robust error handling: null pointer checks, dimension mismatch detection, allocation failure paths</li>
<li>Add thread-safety annotations and document re-entrancy guarantees</li>
<li>Implement a simple runtime CPU feature detection stub (for future multi-target dispatch)</li>
<li>Write a comprehensive <code>API.md</code> reference document</li>
</ul>
<p><strong>Deliverables:</strong> Stable <code>rv_sparse.h</code>, <code>API.md</code>, error-handling test suite.</p>
<hr>
<h4>Week 9: Benchmark Suite</h4>
<p><strong>Goals:</strong> Build a rigorous, reproducible benchmark framework.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Implement <code>benchmarks/</code> framework measuring: GFLOPS/s, memory bandwidth, latency per operation</li>
<li>Benchmark against scalar baselines across matrix sizes: 128×128, 512×512, 2048×2048, 8192×8192</li>
<li>Benchmark across sparsity levels: 0.1%, 1%, 5%, 10%, 25%, 50%</li>
<li>Benchmark both fp32 and fp64 precision</li>
<li>Add matrix market (<code>.mtx</code>) file reader to benchmark on real-world sparse matrices (e.g., from SuiteSparse collection)</li>
<li>Generate reproducible benchmark reports in CSV and Markdown</li>
</ul>
<p><strong>Deliverables:</strong> <code>benchmarks/</code> suite, benchmark report with graphs, real-world matrix test results.</p>
<hr>
<h4>Week 10: Testing, CI Hardening, and Fuzzing</h4>
<p><strong>Goals:</strong> Ensure correctness and robustness under all conditions.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Expand unit test coverage to 100% of public API surface</li>
<li>Add property-based / fuzz tests: randomized matrix dimensions, values, sparsity patterns</li>
<li>Test edge cases: empty matrices, single-element matrices, integer overflow in index arithmetic, very high sparsity (&gt;99.9%)</li>
<li>Integrate address sanitizer and UBSan into CI builds</li>
<li>Add a valgrind memcheck CI step for memory leak detection</li>
</ul>
<p><strong>Deliverables:</strong> Comprehensive test suite, CI passing with sanitizers, zero memory leaks.</p>
<hr>
<h3>Phase 4 — Integration, Examples, and Final Delivery (Weeks 11–12)</h3>
<h4>Week 11: Example Applications and Integration Guide</h4>
<p><strong>Goals:</strong> Demonstrate practical utility and ease of integration.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Write three example programs in <code>examples/</code>:
<ol>
<li><strong>PageRank kernel:</strong> SpMV iteration on a sparse graph adjacency matrix</li>
<li><strong>Pruned neural network inference:</strong> SpMM for a sparse weight layer forward pass</li>
<li><strong>Conjugate Gradient solver:</strong> SpMV-based iterative linear solver</li>
</ol>
</li>
<li>Write <code>INTEGRATION.md</code>: how to link <code>rv-sparse</code> into an existing C/C++ project with CMake</li>
<li>Provide a pkg-config file for easy discovery</li>
<li>Write a blog-post-style technical writeup explaining the vectorization approach for new contributors</li>
</ul>
<p><strong>Deliverables:</strong> Three working examples, <code>INTEGRATION.md</code>, <code>rv-sparse.pc</code> pkg-config file.</p>
<hr>
<h4>Week 12: Final Polish, Release, and Handoff</h4>
<p><strong>Goals:</strong> Deliver a complete, well-documented open-source library ready for community use.</p>
<p><strong>Tasks:</strong></p>
<ul>
<li>Address all mentor feedback from Weeks 1–11</li>
<li>Write the final <code>README.md</code>: project overview, build instructions, quick-start example, benchmark summary, API overview</li>
<li>Tag <code>v0.1.0</code> release with changelog</li>
<li>Prepare and submit final LFX mentorship report with: implemented features, benchmark results, lessons learned, and suggested future work (e.g., SpTRSV, SpLU, multi-threading via OpenMP)</li>
<li>Open issues for future contributors covering planned extensions</li>
</ul>
<p><strong>Deliverables:</strong> <code>v0.1.0</code> tagged release, complete documentation, final LFX report.</p>
<hr>
<h2>Summary Timeline</h2>

Week | Phase | Key Deliverable
-- | -- | --
1 | Foundation | Build system, CI, repo bootstrap
2 | Foundation | CSR/CSC format implementation
3 | Foundation | Scalar SpMV and SpMM baselines
4 | Vectorization | RVV CSR SpMV kernel
5 | Vectorization | RVV CSC SpMV + masked variants
6 | Vectorization | RVV SpMM core implementation
7 | Vectorization | SpMM blocking + memory optimization
8 | Hardening | Final API + error handling
9 | Benchmarking | Full benchmark suite + real-world matrices
10 | Testing | Fuzz testing + sanitizer CI
11 | Integration | Examples + integration guide
12 | Delivery | v0.1.0 release + final report

<hr>
</body></html>
